### ISSUE TITLE
invalid-argument-type false positives with os.walk/fnmatch code

### ISSUE THREAD

**Issue by correctmost** [astral-sh/ty#1671]
### Summary

After https://github.com/astral-sh/ruff/pull/21553 landed, I started seeing `invalid-argument-type` errors with this code:

```python
import fnmatch
import os
from pathlib import Path


for directory, _, files in os.walk(Path('.')):
    for filename in fnmatch.filter(files, '*.json'):
        continue
```

```
walk.py:7:36: error[invalid-argument-type] Argument to function `filter` is incorrect: Argument type `Path` does not satisfy constraints (`str`, `bytes`) of type variable `AnyStr`
walk.py:7:36: error[invalid-argument-type] Argument to function `filter` is incorrect: Expected `Iterable[str]`, found `list[Path]`
```

Pyright, Pyrefly, and mypy do not emit similar warnings.

### Version

ty 0.0.1-alpha.29

**Comment by sharkdp** [comment]
Thank you for reporting this.

https://github.com/astral-sh/ruff/pull/21553 allowed us to understand generic type aliases, so we now understand [typeshed's `GenericPath`](https://github.com/python/typeshed/blob/148b8e631d9a76f7bdcc5478d36692ff79e16227/stdlib/_typeshed/__init__.pyi#L184) which is used in the signature of `os.walk`:
```py
AnyStr = TypeVar("AnyStr", str, bytes)
GenericPath: TypeAlias = AnyStr | PathLike[AnyStr]

def walk(
    top: GenericPath[AnyStr], topdown: bool = True, onerror: _OnError | None = None, followlinks: bool = False
) -> Iterator[tuple[AnyStr, list[AnyStr], list[AnyStr]]]:
	…
```

Since we can't handle generic protocols in the generic solver yet, we solve `AnyStr` as `Path` here (first union element), instead of `str` (second union element), and therefore return `list[Path]` in the third tuple element instead of `list[str]`.

This should be solved by #623.

**Comment by correctmost** [comment]
It looks like this issue was fixed by astral-sh/ruff@7bf50e70.

https://play.ty.dev/15dcb62d-3910-42f7-8121-543ebfd3fb89

**Comment by AlexWaygood** [comment]
Aha! So it looks like this problem was caused by two bugs existing in ty simultaneously. Fixing either would have fixed this symptom, and one has now been fixed (by https://github.com/astral-sh/ruff/commit/7bf50e70)

**Comment by AlexWaygood** [comment]
Thanks @correctmost!
